CMake handling of missing LLD could be improved

Hi, we already talked about this shortly after today's seminar session: On Arch Linux (and Arch based distros such as Manjaro), `cmake` fails to generate the build files for hyrise. While the first error is usually some missing library (regex, thread), one can see that some basic compiler flag checks (for example whether `-Werror` works) also fail, so it seems like the entire toolchain is somehow misconfigured. For repro, the following Dockerfile can be used:

```dockerfile
FROM archlinux:base-devel-20230604.0.155602

RUN pacman -Syu --noconfirm && pacman -S --noconfirm git clang make ncurses boost cmake python

RUN  useradd --create-home --shell /bin/bash hyrise
USER hyrise
WORKDIR /home/hyrise

RUN git clone --recurse-submodules --shallow-submodules https://github.com/hyrise/hyrise
WORKDIR /home/hyrise/hyrise

RUN mkdir build && cd build && cmake -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ ..
```

Note that I used clang as a compiler because the installed GCC is GCC 13 which I wanted to rule out as failing for just being too new and unstable - the same errors happen with it though.

The output of the `cmake` command is
```plain
-- The C compiler identification is Clang 15.0.7
-- The CXX compiler identification is Clang 15.0.7
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/sbin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/sbin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test COMPILER_SUPPORTS_MARCH_NATIVE
-- Performing Test COMPILER_SUPPORTS_MARCH_NATIVE - Success
-- WARNING: Numa library not found.
-- Try: 'sudo apt-get install libnuma libnuma-dev'
-- WARNING: tbb library not found.
-- Try: 'sudo apt-get install libtbb-dev'
-- Found readline library: inc=/usr/include, lib=/usr/lib/libreadline.so
-- Looking for wsyncup in /usr/lib/libcurses.so
-- Looking for wsyncup in /usr/lib/libcurses.so - not found
-- Looking for wsyncup in /usr/lib/libncurses.so
-- Looking for wsyncup in /usr/lib/libncurses.so - not found
-- Looking for cbreak in /usr/lib/libncurses.so
-- Looking for cbreak in /usr/lib/libncurses.so - not found
-- Looking for nodelay in /usr/lib/libncurses.so
-- Looking for nodelay in /usr/lib/libncurses.so - not found
-- Found Curses: /usr/lib/libncurses.so
-- Found SQLite3: /usr/include (found version "3.42.0")
-- Found Boost: /usr/include (found version "1.81.0") found components: container date_time
-- Failed to find LLVM FileCheck
-- Found Git: /usr/sbin/git (found version "2.41.0")
-- git version: v1.6.1 normalized to 1.6.1
-- Version: 1.6.1
-- Looking for shm_open in rt
-- Looking for shm_open in rt - not found
-- Performing Test HAVE_CXX_FLAG_STD_CXX11
-- Performing Test HAVE_CXX_FLAG_STD_CXX11 - Failed
-- Performing Test HAVE_CXX_FLAG_STD_CXX0X
-- Performing Test HAVE_CXX_FLAG_STD_CXX0X - Failed
-- Performing Test HAVE_CXX_FLAG_WALL
-- Performing Test HAVE_CXX_FLAG_WALL - Failed
-- Performing Test HAVE_CXX_FLAG_WEXTRA
-- Performing Test HAVE_CXX_FLAG_WEXTRA - Failed
-- Performing Test HAVE_CXX_FLAG_WSHADOW
-- Performing Test HAVE_CXX_FLAG_WSHADOW - Failed
-- Performing Test HAVE_CXX_FLAG_WERROR
-- Performing Test HAVE_CXX_FLAG_WERROR - Failed
-- Performing Test HAVE_CXX_FLAG_PEDANTIC
-- Performing Test HAVE_CXX_FLAG_PEDANTIC - Failed
-- Performing Test HAVE_CXX_FLAG_PEDANTIC_ERRORS
-- Performing Test HAVE_CXX_FLAG_PEDANTIC_ERRORS - Failed
-- Performing Test HAVE_CXX_FLAG_WSHORTEN_64_TO_32
-- Performing Test HAVE_CXX_FLAG_WSHORTEN_64_TO_32 - Failed
-- Performing Test HAVE_CXX_FLAG_FSTRICT_ALIASING
-- Performing Test HAVE_CXX_FLAG_FSTRICT_ALIASING - Failed
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED_DECLARATIONS
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED_DECLARATIONS - Failed
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED - Failed
-- Performing Test HAVE_CXX_FLAG_WD654
-- Performing Test HAVE_CXX_FLAG_WD654 - Failed
-- Performing Test HAVE_CXX_FLAG_WTHREAD_SAFETY
-- Performing Test HAVE_CXX_FLAG_WTHREAD_SAFETY - Failed
-- Performing Test HAVE_CXX_FLAG_COVERAGE
-- Performing Test HAVE_CXX_FLAG_COVERAGE - Failed
-- Performing Test HAVE_STD_REGEX
-- Performing Test HAVE_STD_REGEX
-- Performing Test HAVE_STD_REGEX -- failed to compile
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Performing Test HAVE_POSIX_REGEX
-- Performing Test HAVE_POSIX_REGEX
CMake Error at third_party/benchmark/CMakeLists.txt:304 (message):
  Failed to determine the source files for the regular expression backend


-- Performing Test HAVE_POSIX_REGEX -- failed to compile
-- Configuring incomplete, errors occurred!
```

As an aside, the following single file CMake project from the current exercise in the course Hardware Conscious Data Processing successfully builds, also including googletest (I can't confirm what happens with the benchmark, because that is only run in CI):

```cmake
cmake_minimum_required(VERSION 3.16)
project(art)
set(CMAKE_CXX_STANDARD 20)

include(FetchContent)

option(CI_BUILD "Set to ON for complete build in CI." OFF)

if(NOT CMAKE_BUILD_TYPE)
    message(STATUS "No build type specified. Defaulting to Debug.
                    Use cmake `-DCMAKE_BUILD_TYPE=Release ..` to build an optimized version.")
endif()

if (CMAKE_BUILD_TYPE MATCHES Release)
    message(STATUS "Building an optimized release version.")
    add_compile_options(-O3)
    if (CI_BUILD)
        # We cannot set this generally, as it does not work on all architectures.
        message(STATUS "Adding -march/tune=native in CI")
        add_compile_options(-march=native -mtune=native)
    endif()
endif()

set(TASK_SOURCES src/art.cpp src/art.hpp src/key.hpp)

add_library(art ${TASK_SOURCES})
target_include_directories(art INTERFACE ${CMAKE_CURRENT_SOURCE_DIR}/src)
# Pass all available SIMD options on our server to GCC. Disable warning if unused.
target_compile_options(art PUBLIC
        -mmmx -msse -msse2 -msse3 -mssse3 -msse4 -msse4a -msse4.1 -msse4.2 -mavx
        -mavx2 -mavx512f -mavx512cd -mavx512vl -mavx512bw -mavx512dq -mavx512vnni)

enable_testing()
FetchContent_Declare(
        googletest
        GIT_REPOSITORY https://github.com/google/googletest.git
        GIT_TAG v1.13.0
)

set(BUILD_GMOCK ON CACHE BOOL "" FORCE)
set(BUILD_GTEST ON CACHE BOOL "" FORCE)
FetchContent_MakeAvailable(googletest)

add_executable(basic_test test/basic.cpp)
add_test(basic_test basic_test)
target_link_libraries(basic_test art gtest gmock)

if (${CI_BUILD})
    # Build advanced tests in CI only
    add_executable(advanced_test test/advanced.cpp)
    add_test(advanced_test advanced_test)
    target_link_libraries(advanced_test art gtest gmock)

    add_executable(hdp_benchmark test/benchmark.cpp)
    target_link_libraries(hdp_benchmark art)
endif()
```

Output:
```plain
-- The C compiler identification is Clang 15.0.7
-- The CXX compiler identification is Clang 15.0.7
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/sbin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/sbin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- No build type specified. Defaulting to Debug.
                    Use cmake `-DCMAKE_BUILD_TYPE=Release ..` to build an optimized version.
-- Found Python: /usr/sbin/python3.11 (found version "3.11.3") found components: Interpreter
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Configuring done (43.8s)
-- Generating done (0.0s)
```

For the seminar, building in an Ubuntu docker container works, but maybe you still want to find the underlying problem in the `cmake` config :)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,5 +1,5 @@
 # consider updating DEPENDENCIES.md when you touch this line
-cmake_minimum_required(VERSION 3.9 FATAL_ERROR)
+cmake_minimum_required(VERSION 3.18 FATAL_ERROR)
 
 project(Hyrise)
 
@@ -105,12 +105,6 @@ set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} ${FLAGS_ALL} ${FLAGS_REL
 set(CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS_RELWITHDEBINFO} ${FLAGS_ALL} ${FLAGS_RELEASE}")
 set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} ${FLAGS_ALL} ${FLAGS_RELEASE}")
 
-# Check for unity builds
-if(${CMAKE_VERSION} VERSION_LESS "3.16.0")
-    if(DEFINED CMAKE_UNITY_BUILD)
-        message(FATAL_ERROR "-DCMAKE_UNITY_BUILD=ON was set, but your cmake version does not support unity builds.")
-    endif()
-endif()
 
 # On Linux, use the lld linker, which is twice as fast for incremental debug builds
 if (NOT APPLE)
@@ -151,17 +145,14 @@ cmake_policy(SET CMP0069 NEW)
 set(CMAKE_POLICY_DEFAULT_CMP0069 NEW)
 option(ENABLE_LTO "Set to ON to build Hyrise with enabled Link Time Optimization (LTO). Default: OFF" OFF)
 if (${ENABLE_LTO})
-    if(NOT ${CMAKE_VERSION} VERSION_LESS "3.9.0")
-        cmake_minimum_required(VERSION 3.9 FATAL_ERROR)
-        include(CheckIPOSupported)
-        check_ipo_supported(RESULT ipo_supported OUTPUT ipo_output)
-        if(ipo_supported)
-            set(CMAKE_INTERPROCEDURAL_OPTIMIZATION TRUE)
-            add_definitions(-DWITH_LTO)
-            message(STATUS "Building with Link-Time Optimization")
-        else()
-            message(WARNING "LTO not supported: ${ipo_output}")
-        endif()
+    include(CheckIPOSupported)
+    check_ipo_supported(RESULT ipo_supported OUTPUT ipo_output)
+    if(ipo_supported)
+        set(CMAKE_INTERPROCEDURAL_OPTIMIZATION TRUE)
+        add_definitions(-DWITH_LTO)
+        message(STATUS "Building with Link-Time Optimization")
+    else()
+        message(WARNING "LTO not supported: ${ipo_output}")
     endif()
 endif()
 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -105,11 +105,25 @@ set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} ${FLAGS_ALL} ${FLAGS_REL
 set(CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS_RELWITHDEBINFO} ${FLAGS_ALL} ${FLAGS_RELEASE}")
 set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} ${FLAGS_ALL} ${FLAGS_RELEASE}")
 
-
-# On Linux, use the lld linker, which is twice as fast for incremental debug builds
-if (NOT APPLE)
-    set(CMAKE_SHARED_LINKER_FLAGS "-fuse-ld=lld")
-    set(CMAKE_EXE_LINKER_FLAGS "-fuse-ld=lld")
+include(CheckLinkerFlag)
+if (HYRISE_LINKER)
+    check_linker_flag(CXX "-fuse-ld=${HYRISE_LINKER}" HAS_HYRISE_LINKER)
+    if (NOT HAS_HYRISE_LINKER)
+        message(FATAL_ERROR "Got HYRISE_LINKER=${HYRISE_LINKER}, but -fuse-ld=${HYRISE_LINKER} does not work.")
+    endif()
+    message(STATUS "Using HYRISE_LINKER (${HYRISE_LINKER}).")
+    add_link_options("-fuse-ld=${HYRISE_LINKER}")
+else()
+    check_linker_flag(CXX "-fuse-ld=lld" HAS_LLD)
+    if (HAS_LLD)
+        message(STATUS "Found LLD, using as linker.")
+        add_link_options("-fuse-ld=lld")
+    else()
+        message(STATUS "Using default linker.")
+        if (NOT APPLE)
+            message(WARNING "Consider installing the LLD linker for faster incremental builds on Linux.")
+        endif()
+    endif()
 endif()
 
 # Require NCurses over Curses

--- a/DEPENDENCIES.md
+++ b/DEPENDENCIES.md
@@ -16,7 +16,7 @@
 | libnuma-dev               | any              |    Linux |                            Yes (numa) |
 | libnuma1                  | any              |    Linux |                            Yes (numa) |
 | libpq-dev                 | >= 9             |    All   |                                    No |
-| lld                       | any              |    Linux |   No, but could be removed from cmake |
+| lld                       | any              |    Linux |                                   Yes |
 | parallel                  | any              |    All   |                                   Yes |
 | pexpect                   | >= 4             |    All   |                     Yes (tests in CI) |
 | postgresql-server-dev-all | >= 154           |    Linux |                                    No |

--- a/DEPENDENCIES.md
+++ b/DEPENDENCIES.md
@@ -8,7 +8,7 @@
 | clang-format              | >= 11.0          |    All   |                      Yes (formatting) |
 | clang-tidy                | >= 11.0          |    All   |                         Yes (linting) |
 | coreutils                 | any              |    Mac   |                         Yes (scripts) |
-| cmake                     | >= 3.9           |    All   |                                    No |
+| cmake                     | >= 3.18          |    All   |                                    No |
 | dos2unix                  | any              |    All   |                         Yes (linting) |
 | gcc                       | >= 9.1           |    All   | Yes, if clang installed, not for OS X |
 | gcovr                     | >= 3.2           |    All   |                        Yes (coverage) |

--- a/src/lib/CMakeLists.txt
+++ b/src/lib/CMakeLists.txt
@@ -754,14 +754,12 @@ target_include_directories(hyrise_impl PUBLIC ${CMAKE_BINARY_DIR})
 if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
 target_compile_options(hyrise_impl PUBLIC -Xclang -fno-pch-timestamp)
 endif()
-if(NOT ${CMAKE_VERSION} VERSION_LESS "3.16.0")
-    target_precompile_headers(hyrise_impl PRIVATE
-      all_parameter_variant.hpp
-      storage/create_iterable_from_segment.hpp
-      storage/table.hpp
-      types.hpp
-    )
-endif()
+target_precompile_headers(hyrise_impl PRIVATE
+  all_parameter_variant.hpp
+  storage/create_iterable_from_segment.hpp
+  storage/table.hpp
+  types.hpp
+)
 
 # -fPIC generates position independent code which is necessary because plugins might access hyrise functions.
 target_compile_options(hyrise_impl PRIVATE -fPIC)

--- a/src/test/CMakeLists.txt
+++ b/src/test/CMakeLists.txt
@@ -294,15 +294,13 @@ target_link_libraries(hyriseTest hyrise ${LIBRARIES})
 if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
 target_compile_options(hyriseTest PUBLIC -Xclang -fno-pch-timestamp)
 endif()
-if(NOT ${CMAKE_VERSION} VERSION_LESS "3.16.0")
-    target_precompile_headers(hyriseTest PRIVATE
-        <gtest/gtest.h>
-        [["all_parameter_variant.hpp"]]
-        [["storage/create_iterable_from_segment.hpp"]]
-        [["storage/table.hpp"]]
-        [["types.hpp"]]
-    )
-endif()
+target_precompile_headers(hyriseTest PRIVATE
+    <gtest/gtest.h>
+    [["all_parameter_variant.hpp"]]
+    [["storage/create_iterable_from_segment.hpp"]]
+    [["storage/table.hpp"]]
+    [["types.hpp"]]
+)
 
 # Configure hyriseSystemTest
 add_executable(hyriseSystemTest ${SYSTEM_TEST_SOURCES})

